packages/matrix-client-server/src/index.test.ts

```diff
@@ -485,6 +519,71 @@ describe('Use configuration file', () => {
           expect(response.body).toHaveProperty('error')
           expect(response.body).toHaveProperty('errcode')
         })
+        it('should refuse autenticating an appservice without a token', async () => {
+          const response = await request(app)
+            .post('/_matrix/client/v3/register')
+            .set('User-Agent', 'curl/7.31.0-DEV')
+            .query({ kind: 'user' })
+            .send({
+              auth: {
+                type: 'm.login.application_service',
+                username: '_irc_bridge_'
+              }
+            })
+          expect(response.statusCode).toBe(401)
+          expect(response.body).toHaveProperty('error')
+          expect(response.body).toHaveProperty('errcode', 'M_MISSING_TOKEN')
+        })
+        it('should refuse authenticating an appservice with the wrong token', async () => {
+          const response = await request(app)
+            .post('/_matrix/client/v3/register')
+            .set('User-Agent', 'curl/7.31.0-DEV')
+            .set('Authorization', `Bearer wrongToken`)
+            .query({ kind: 'user' })
+            .send({
+              auth: {
+                type: 'm.login.application_service',
+                username: '_irc_bridge_'
+              }
+            })
+          expect(response.statusCode).toBe(401)
+          expect(response.body).toHaveProperty('error')
+          expect(response.body).toHaveProperty('errcode', 'M_UNKNOWN_TOKEN')
+        })
+        it('should refuse authenticating an appservice with a username that is too long', async () => {
+          const asToken = conf.application_services[0].as_token
+          const response = await request(app)
+            .post('/_matrix/client/v3/register')
+            .set('User-Agent', 'curl/7.31.0-DEV')
+            .set('Authorization', `Bearer ${asToken}`)
+            .query({ kind: 'user' })
+            .send({
+              auth: {
+                type: 'm.login.application_service',
+                username: 'invalidUser'
+              }
+            })
+          expect(response.statusCode).toBe(401)
+          expect(response.body).toHaveProperty('error')
+          expect(response.body).toHaveProperty('errcode', 'M_INVALID_USERNAME')
+        })
+        it('should refuse authenticating an appservice with a username it has not registered', async () => {
```

Comment:
Does this test prove that a normal user token can not be used as app token ?

Reply:
No, I did not test that feature but my authentication function only checks if the token corresponds to an appservice token after checking that it is not a known user token. I can add a test for this if necessary

Reply:
Can you add such test to prove security ?

Reply:
Yes, will do this right away